Support field `fileData` (direct file URL) for `GeminiModel` and `GoogleModel`

vricciardulli opened the pull request:
- Fixes https://github.com/pydantic/pydantic-ai/issues/1134

Using the existing `{Document,Audio,Image}Url` and having `GeminiModel` and `GoogleModel` determine whether it can send them using `FileData.fileUri` instead of downloading them.

Added support for Youtube URLs and `gs://` URLs as well

Per provider:
* GLA: only Youtube URLs are sent directly, for the rest, data is downloaded and sent as base64-encoded string.
* VertexAI: By default, all URLs are sent directly and hence no data is downloaded. However, the user can set flag `force_download` to `True` to force pydantic-ai to download and send the file content instead. This is useful to circumvent crawling restrictions.

vricciardulli commented:
Also allowed `FileUrl` to be passed to GLA-Gemini models, as it is supported as the [docs show](https://ai.google.dev/api/caching#FileData). Although that documentation does not provide much information and PDF files don't seem to be supported. YouTube URLs seem to be supported by the GLA.

ioga commented:
Thanks a lot for making this PR, it worked for me. Would love to see it get merged. Couple suggestions:
1. Since youtube videos is the only supported type, I think the example should use them instead of the pdf URLs which do not work.
2. `media_type` field is not required for youtube URLs to work, it should be made optional.

vricciardulli commented:
> Thanks a lot for making this PR, it worked for me. Would love to see it get merged. Couple suggestions:
> 
> 1. Since youtube videos is the only supported type, I think the example should use them instead of the pdf URLs which do not work.
> 2. `media_type` field is not required for youtube URLs to work, it should be made optional.

Thanks for the interest!

Did you try this with both GLA and Vertex providers? because PDF URLs seem to work with Vertex. I agree that with GLA only youtub URLs seem to work though.

ioga commented:
Can you just add both youtube and PDF examples? 

My second point, making media type optional, is probably more material: right now Gemini seems to ignore it and you can just pass whatever, but no way to know if google will change that behavior in the future.

vricciardulli commented:
@ioga added, let me know what you think. Still need to increase test coverage and maybe for the Vertex provider the mime type should be inferred when possible, because it is required for that provider.

ioga commented:
works for me on `google-gla`, thank you.

kraft87 commented:
Any update on this?

DouweM commented:
@vricciardulli What do you think about using the existing `{Document,Audio,Image}Url` for this, and having `GeminiModel` determine whether it can send them using `FileData.fileUri` instead of doing the download? We already do that in `AnthropicModel` for `ImageUrl` and `DocumentUrl` for example.

The YouTube URL could similarly be implemented on the existing `VideoUrl`, with an error raised from models that don't support it as downloading the URL wouldn't work.

vricciardulli commented:
> @vricciardulli What do you think about using the existing `{Document,Audio,Image}Url` for this, and having `GeminiModel` determine whether it can send them using `FileData.fileUri` instead of doing the download? We already do that in `AnthropicModel` for `ImageUrl` and `DocumentUrl` for example.
> 
> The YouTube URL could similarly be implemented on the existing `VideoUrl`, with an error raised from models that don't support it as downloading the URL wouldn't work.

@DouweM thanks for the feedback. let me give it a try

dhimmel commented on tests/models/test_gemini_vertex.py:
```diff
@@ -0,0 +1,119 @@
+            ImageUrl(url='https://upload.wikimedia.org/wikipedia/commons/6/6a/Www.wikipedia_screenshot_%282021%29.png'),
+            id='ImageUrl',
+        ),
+        pytest.param(ImageUrl(url='gs://my-bucket/path/to/image-blob.png'), id='ImageUrl (gs://)'),
```
Awesome to see that this PR intends to support GCS URIs like `gs://cloud-samples-data/generative-ai/image/scones.jpg`!

DouweM commented on pydantic_ai_slim/pydantic_ai/models/gemini.py:
```diff
@@ -342,15 +343,25 @@ async def _map_user_prompt(self, part: UserPromptPart) -> list[_GeminiPartUnion]
                         _GeminiInlineDataPart(inline_data={'data': base64_encoded, 'mime_type': item.media_type})
                     )
+                elif isinstance(item, VideoUrl) and item.is_youtube:
+                    file_data = _GeminiFileDataPart(file_data={'file_uri': item.url, 'mime_type': item.media_type})
```
Just to confirm, this works with both google-gla and google-vertex?

vricciardulli replied:
Yes, correct: YouTube URLs work with both providers.

However, there's a limit of 1 YouTube URL per request. If more are provided, Gemini returns 400 and body:
```json
{
  "error": {
    "code": 400,
    "message": "You can only include 1 youtube link. Please reduce the number accordingly.",
    "status": "INVALID_ARGUMENT"
  }
}
```

Should I enforce this limit in `pydantic-ai`, or should the user deal with it?

DouweM commented on pydantic_ai_slim/pydantic_ai/models/gemini.py:
```diff
@@ -342,15 +343,25 @@ async def _map_user_prompt(self, part: UserPromptPart) -> list[_GeminiPartUnion]
-                    )
-                    content.append(inline_data)
+                    if provider == 'google-vertex':
+                        file_data = _GeminiFileDataPart(file_data={'file_uri': item.url, 'mime_type': item.media_type})
```
This works for all of the above file types, meaning `_GeminiInlineDataPart` is never needed on google-vertex? And similarly, `_GeminiFileDataPart` never works on google-gla so we can't use this logic there?

vricciardulli replied:
For google-GLA: the [documentation doesn't say much](https://ai.google.dev/api/caching#FileData), unfortunately. By testing locally, I could only get it to work with YouTube URLs. So, apart from that single use-case, we never use `_GeminiFileDataPart` for GLA.

For google-Vertex:
- There are different limitations (for example on the size) when using inline-data ([docs](https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/inference#blob)) vs. file-data ([docs](https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/inference#filedata)).
- Maybe we could add a flag to the `*Url` classes `force_download` (`False` by default), so that the user can decide whether or not to download and send base64-encoded data? It does make things more complex though and not sure there's much gain

DouweM commented on tests/models/test_gemini.py:
```diff
@@ -1021,6 +1022,83 @@ async def get_image() -> BinaryContent:
+    env: TestEnv,
+    allow_model_requests: None,
+) -> None:
+    def handler(request: httpx.Request) -> httpx.Response:
```
Instead of directly inspecting these requests and faking the responses, can we use `pytest.mark.vcr` to actually perform these requests once and record them, and then verify that the model output is as expected?

vricciardulli replied:
@DouweM do you have any tips on how to deal with VertexAI and cassettes? the cassettes I uploaded won't work because the CI is not properly configured for Vertex AI as far as I understand

The google-GLA cassette works instead.

DouweM reviewed (changes requested):
@vricciardulli Thank you! In addition to the changes I've requested, can you please update the https://ai.pydantic.dev/input/ documentation to mention these URLs are supported on those specific models?

It'd also be good to raise an error on other models that don't support these URLs. What do you think of moving the download logic that's common among many models to a new helper function, that raises an error when `VideoUrl.is_youtube` or any URL `is_gs`?

vricciardulli commented on pydantic_ai_slim/pydantic_ai/models/gemini.py:
```diff
@@ -333,6 +333,7 @@ async def _map_user_prompt(self, part: UserPromptPart) -> list[_GeminiPartUnion]
         if isinstance(part.content, str):
             return [{'text': part.content}]
         else:
+            provider = self._provider.name if isinstance(self._provider, Provider) else self._provider
```
changed in https://github.com/pydantic/pydantic-ai/pull/1136/commits/0bf27ed23ce02989409efc7b87f23c05aa8699c5

DouweM commented:
@vricciardulli There's a big refactor of the Google providers and model happening in https://github.com/pydantic/pydantic-ai/pull/1373, and merging this first would result in merge conflicts that'd be hard to resolve on that side -- it'll be easier to reimplement this once that lands. So I'm going to put this on hold for the moment and we can rebase, update, and look at it again once that's done, OK?

vricciardulli commented on tests/models/test_gemini_vertexai.py:
Content of this file are not in `test_gemini_vertex.py`

vricciardulli commented:
@DouweM I think this is now ready for review.

Let me summarize what was done:
1. By default, for `GeminiModel` and `GoogleModel`, file URLs are sent directly in the API request. No download is performed.
2. However, a user can decide to force the download by setting the Boolean flag `force_download` to `True` (implemented by using an ABC in commit https://github.com/pydantic/pydantic-ai/pull/1136/commits/0514fe822ede000815c0d2727dfa799f4e09487e). This is needed because Gemini can't crawl all websites.
3. I used your implementation of `download_item` from https://github.com/pydantic/pydantic-ai/pull/1613/files to abstract common download logic and raise errors when `gs://` or YouTube URLs are provided.
4. Updated the docs with an explanation of what goes on behind the scenes.
5. Added tests and used VCR where needed.

**Unresolved issues**:
1. For some reason that I can't figure out, parts of function `download_item` are still uncovered, even though I've added tests for that part specifically (as far as I understand, it's the part of code that handle contents of type `application/octet-stream`)

vricciardulli commented:
@DouweM thank you for the review! as per your comments:
1. I fixed the docs and most other comments in https://github.com/pydantic/pydantic-ai/pull/1136/commits/209522ed7098a2985c5c5e9fac39899051aba32a
2. I omitted the check for `force_download` when `item` is a YouTube URL and removed the associated tests in https://github.com/pydantic/pydantic-ai/pull/1136/commits/c3b2993be5fdab5d1424339d18807700780624c1

dhimmel commented on docs/input.md:
````diff
@@ -109,3 +102,25 @@ result = agent.run_sync(
 # > The document discusses...
 ```
+
+## User-side download vs. direct file URL
````
Great docs in this section and sensible design!

DouweM commented:
@vricciardulli Thanks Vittorio! I've approved.

@Kludex Can you please have a final look as well?

DouweM commented:
@vricciardulli Note that there's a merge conflict!

Kludex commented:
I'd like to review this. Please don't merge it.

vricciardulli commented:
Conflicts fixed, happy to hear your thoughts as well when you have time @Kludex

dhimmel commented:
> happy to hear your thoughts as well when you have time @Kludex

Pinging @Kludex regarding the pending review here as an interested user of `gs://` URIs and someone looking to save @vricciardulli from having to reconcile more conflicts.

Kludex commented on pydantic_ai_slim/pydantic_ai/models/__init__.py:
```diff
@@ -611,6 +613,79 @@ def _cached_async_http_transport() -> httpx.AsyncHTTPTransport:
+    item: FileUrl,
+    data_format: Literal['bytes'],
+    type_format: Literal['mime', 'extension'] = 'mime',
+) -> tuple[bytes, str]:
+    pass
```
```suggestion
) -> tuple[bytes, str]: ...
```

Kludex commented on pydantic_ai_slim/pydantic_ai/models/__init__.py:
```diff
@@ -611,6 +613,79 @@ def _cached_async_http_transport() -> httpx.AsyncHTTPTransport:
+    item: FileUrl,
+    data_format: Literal['base64', 'base64_uri', 'text'],
+    type_format: Literal['mime', 'extension'] = 'mime',
+) -> tuple[str, str]:
```
Instead of `tuple[str, str]`, can we have a `DownloadedItem` `TypedDict` instead?

vricciardulli commented on docs/input.md:
```diff
@@ -109,3 +102,25 @@ result = agent.run_sync(
+
+    See the [Gemini API docs for Vertex AI](https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/inference#filedata) to learn more about supported URLs, formats and limitations:
+
+    - Cloud Storage bucket URIs (with protocol `gs://`)
+
+    - Public HTTP(S) URLs
+
+    - Public YouTube video URL (maximum one URL per request)
```
It looks like this:

<img width="652" alt="Screenshot" src="https://github.com/user-attachments/assets/4aed7597-dde9-4d0a-9000-78dd2b1014f8" />

Kludex replied:
```suggestion
    - Cloud Storage bucket URIs (with protocol `gs://`)
    - Public HTTP(S) URLs
    - Public YouTube video URL (maximum one URL per request)
```

vricciardulli commented on docs/input.md:
```diff
@@ -109,3 +102,25 @@ result = agent.run_sync(
+
+    - Public YouTube video URL (maximum one URL per request)
+
+    However, because of crawling restrictions, it may happen that Gemini can't access certain URLs. In that case, you can instruct PydanticAI to download the file content and send that instead of the URL by setting the boolean flag `force_download` to `True`. This attribute is available on all objects that inherit from [`FileUrl`][pydantic_ai.messages.FileUrl].
```
Same as above, it's to have it aligned with the second bullet point, but on a new line after the list of supported URLs (check the image in the first comment please)

Kludex reviewed (approved):
Sorry the delay.

Amazing work! :)
